pandas/core/internals.py

```diff
@@ -314,6 +314,15 @@ def ftype(self):
     def merge(self, other):
         return _merge_blocks([self, other])
 
+    def concat_same_type(self, to_concat, placement=None):
```

Comment:
actually you could combine all of these ``concat_same_type`` into a single routine  if you did this

```
def concat_same_type(self, to_concat, placement=None)
       """
       Concatenate list of single blocks of the same type.
       """
        values = self._concatenator([blk.values for blk in to_concat], axis=self.ndim - 1)
        return self.make_block_same_class(
            values, placement=placement or slice(0, len(values), 1))
```

Then add to Block
```
_concatenator = np.concatenate
```
Categorical
```
_concatnator = _concat._concat_categorical
```

etc

Reply:
Ah, that would actually be nice. The only problem with this is that (for the reasons that I had the if/else statements originally) `self.make_block_same_class` will not always work. Eg ``_concat_categorical`` can return both categorical values as object values, and depending on that should return a CategoricalBlock or another type of Block.

Reply:
so for categorical block override; otherwise u end up repeating lots of code

Reply:
but need to overwrite for Datetimetz as well, so then end up with almost as many overridden ones as now (only for Sparse it would not be needed then).

Reply:
this is still repeatting way too much code. you can just do it this way

in Block
```
def concat_same_type(self, to_concat, constructor=None, placement=None):
     values = self._concatenator(......)
    if constructor is None:
        constructor = make_block
      return constructor(....)
```

then where needed

```
def concat_same_type(.......):
      return super(Categorical, self).concat_same_type(....., constructor=self.make_block_same_class)
```

that way for an overriden class you are not repeating evertyhing.